Use expression-based root mapping

naoyam opened the pull request:
Currently, root mapping is done only for direct producer-consumer pairs with implicit assumptions on IterDomain ordering. Now that we have the capability of expression-based mapping (#393), it could be made more robust.

TODOs
- [x] Remove the extra check with the existing root mapping in replayPasC and replayCasP
- [x] Add more comments

Commit:
WIP

Commit:
update

Commit:
Root mapping

Commit:
WIP

csarofeen commented on torch/csrc/jit/codegen/cuda/ir_nodes.cpp:
```diff
@@ -16,9 +16,9 @@ namespace cuda {
 
 namespace {
 
-class ScalarCheck : OptInDispatch {
```
Is this necessary in this PR or can it be split out?

naoyam replied:
Split out. See #429.

csarofeen commented on torch/csrc/jit/codegen/cuda/ir_nodes.cpp:
```diff
@@ -187,8 +187,8 @@ bool TernaryOp::sameAs(const TernaryOp* other) const {
-BroadcastOp::BroadcastOp(Val* _out, Val* _in)
-    : Expr(ExprType::BroadcastOp), out_(_out), in_(_in) {
+BroadcastOp::BroadcastOp(Val* _out, Val* _in, const std::vector<bool>& is_broadcast_dim)
+    : Expr(ExprType::BroadcastOp), out_(_out), in_(_in), is_broadcast_dim_(is_broadcast_dim) {
```
Can adding `is_broadcast_dim_` also be split out of this PR?

naoyam replied:
Will do.

csarofeen commented on torch/csrc/jit/codegen/cuda/kernel_ir.cpp:
```diff
@@ -408,7 +408,7 @@ Allocate::Allocate(
     }
   }
-
+#if 0
```
TODO: Remove this `#if 0`

csarofeen commented on torch/csrc/jit/codegen/cuda/ir_internal_nodes.h:
```diff
@@ -133,11 +133,20 @@ class TORCH_CUDA_API BroadcastOp : public Expr {
     return in_;
   }
 
+  bool isBroadcastDim(size_t dim) const {
```
Add comment that this is relative to the output tensor.

csarofeen commented on torch/csrc/jit/codegen/cuda/ir_internal_nodes.h:
```diff
@@ -133,11 +133,20 @@ class TORCH_CUDA_API BroadcastOp : public Expr {
+    return is_broadcast_dim_.at(dim);
+  }
+
+  const std::vector<bool> getBroadcastDimFlags() const {
```
Add comment that this is relative to the output tensor.

naoyam replied:
Please see #430

csarofeen commented on torch/csrc/jit/codegen/cuda/ir_internal_nodes.h:
```diff
@@ -259,10 +268,7 @@ class TORCH_CUDA_API IterDomain : public Val {
-
-  // Attempt to prove 2 IterDomains are equal in start and rawExtent
-  static bool proveEquivalent(IterDomain* a, IterDomain* b);
+  static const IterDomain* concretizeDomain(const IterDomain* bcast_dom);
```
Add comment as to what concretize domain means.

csarofeen commented on torch/csrc/jit/codegen/cuda/ir_nodes.cpp:
```diff
@@ -1033,118 +1052,6 @@ std::pair<TensorDomain*, TensorDomain*> TensorDomain::rFactor(
-  int next_index_ = 0;
-};
-
 //! Concretize broadcast axes, i.e. identifying a non-broadcast
```
This comment should likely be the comment on IterDomain::concretizeDomain

csarofeen commented on torch/csrc/jit/codegen/cuda/ir_nodes.cpp:
```diff
@@ -1174,12 +1081,12 @@ class ConcretizeDomain : private BackwardVisitor {
 
   // Returns the concretized id recorded from traversal
-  IterDomain* concretized(IterDomain* id) const {
+  const IterDomain* concretized(const IterDomain* id) const {
```
Can topologies exist where a broadcast dimension is concretized to more than one IterDomain's?

e.g.
```.cpp
auto tv0 = makeDummyTensor(1);
fusion.addInput(tv0);
auto tv1 = broadcast(tv0, {false, true};
auto tv2 = makeDummyTensor(2);
fusion.addInput(tv2);
auto tv3 = mul(tv1, tv2);
fusion.addOutput(tv3);
auto tv4 = makeDummyTensor(2);
fusion.addInput(tv4);
auto tv5 = mul(tv1, tv4);
fusion.addOutput(tv5);
```
It seems to me in this case the root domain of `tv4->getRootDomain()[0]` cannot be proven to be equivalent to `tv5->getRootDomain()[0]` and `tv1->getRootDomain()[0]` should be mapped to both.

naoyam replied:
I guess you're asking about `tv1->getRootDomain()[1]` since the 0-th rood domain is not broadcast. In any case, `concretizeDomain` only returns a corresponding `IterDomain` that has a concrete size. Whether that identified concrete `IterDomain` has the same size as other `IterDomain`s is a different question. Pinging @shmsong as this was originally written by him.

shmsong replied:
The assumption of "concretizing" a broadcast axes is "single use"/"single semantic" for the broadcast axes, and if the same broadcast axis is used in multiple places like in the example:
```cpp
auto tv3 = mul(tv1, tv2);
auto tv5 = mul(tv1, tv4);
```
and say
```
tv1 = T1[i0,b1]
tv2 = T2[i1,i2]
tv4 = T4[i3,i4]
```
Then i2 and i4 will be assumed equal, and actually the `IterDomain:: proveEquivalent` will return them as equal. 

One possible way around if tv2 and tv4 don't actually have the same dimensions is explicitly create a new broadcast for each use since they're actually different broadcast semantics (i.e. sizes).

Would be happy to refactor any part if this isn't the particular behavior that we want.

naoyam replied:
I assumed that a single broadcast IterDomain must be uniquely mapped to a single IterDomain with a concrete size. That's likely the case most of the times, but not necessarily true. We should remove the assumption or fail if not.

Having a new broadcast would make things simpler, but that'd be a user-interface question. I think what should be done anyway is that we should be able to detect broadcast IterDomains that may be mapped to multiple different domains.

shmsong replied:
Sure, detection of a `broadcast` mapping to multiple concrete `IterDomains` can be done on top of the `proveEquivalent` class. I can add that somewhere if we want. Meanwhile I was curious how this should be handled in a global root domain mapping, since multiple broadcast semantics will make the result of mapping unavoidably context-dependent.

csarofeen commented on torch/csrc/jit/codegen/cuda/root_domain_map.h:
```diff
@@ -0,0 +1,159 @@
+//! equalities , e.g. :
+//!    T2 [i0,i1] = T1[i2,i3] + T0[i4,i5]
+//! will prove that i2 and i4 are equal in the sense that
+//!    i2.start = i4.start, i2.extent = i4.extent
```
We don't really use at this point, but uncertain why start would have to be the same.

csarofeen commented on torch/csrc/jit/codegen/cuda/root_domain_map.cpp:
```diff
@@ -0,0 +1,379 @@
+namespace {
+
+bool hasMatchingDomains(const std::vector<DomainKey>& domains,
+                        const DisjointSet<DomainKey, DomainKeyHash>& eq_set) {
```
Are these guaranteed to be disjoint sets?

csarofeen commented on torch/csrc/jit/codegen/cuda/root_domain_map.h:
```diff
@@ -0,0 +1,159 @@
+  }
+};
+
+class TORCH_CUDA_API FindIncompatibleDomains : private IterVisitor {
```
Please comment. It seems what you're looking for is any consumer of a reduction cannot be mapped to any iteration domain before that reduction that was reduced. You're having a hard mapping break at reduction IterDomains.

csarofeen commented on torch/csrc/jit/codegen/cuda/root_domain_map.cpp:
```diff
@@ -0,0 +1,379 @@
+  return false;
+}
+
+bool safeToJoin(const std::unordered_set<DomainKey, DomainKeyHash>& domains,
```
Safe to join simply means there is no mapping of the iteration domains across a reduction.

naoyam replied:
It should also detect mappings to multiple axes of a single root domain, which means it needs to be computed at those multiple loop bodies. I'll document what the function does, but `hasMatchingDomains` is the one that should find such cases.

csarofeen commented on torch/csrc/jit/codegen/cuda/root_domain_map.cpp:
```diff
@@ -0,0 +1,379 @@
+  traverse(fusion);
+}
+
+void FindIncompatibleDomains::handle(ReductionOp* op) {
```
Map any reduction IterDomains as incompatible with all IterDomains in all consumer domains.

csarofeen commented on torch/csrc/jit/codegen/cuda/root_domain_map.cpp:
```diff
@@ -0,0 +1,379 @@
+  }
+}
+
+bool FindIncompatibleDomains::isReductionOutputMerged(
```
Returns if reduction iter domain is mapped to a consumer iter domain.

csarofeen commented on torch/csrc/jit/codegen/cuda/root_domain_map.h:
```diff
@@ -0,0 +1,159 @@
+                        const TensorDomain* consumer_td, const IterDomain* consumer_id);
+
+  // Inspect a pointwise or reduction op and record the identified equality
+  void provePointwiseOrReductionOp(Expr* e);
```
Map pointwise IterDomains from inputs of expressions to outputs. Do not map reduction IterDomains in inputs.

csarofeen commented on torch/csrc/jit/codegen/cuda/root_domain_map.h:
```diff
@@ -0,0 +1,159 @@
+
+ private:
+  DisjointSet<DomainKey, DomainKeyHash> eq_set_;
+  std::unordered_map<DomainKey, std::unordered_set<DomainKey, DomainKeyHash>, DomainKeyHash> pending_map_;
```
Keep track of what we want to try and map. Set in `attemptToProveId`

csarofeen commented on torch/csrc/jit/codegen/cuda/root_domain_map.cpp:
```diff
@@ -0,0 +1,379 @@
+  return false;
+}
+
+RootDomainMap::RootDomainMap() {
```
Traverse the entire fusion, setup disjoint sets.

csarofeen commented on torch/csrc/jit/codegen/cuda/root_domain_map.cpp:
```diff
@@ -0,0 +1,379 @@
+  traverseFrom(fusion, leaves, false);
+  //std::cerr << "Traversal done\n";
+  //print(std::cerr);
+  if (!pending_map_.empty()) {
```
Traversal of `Expr`s adds entries into `pending_map_`, visiting `TensorView`s will process those pending entries.

csarofeen commented on torch/csrc/jit/codegen/cuda/root_domain_map.cpp:
```diff
@@ -0,0 +1,379 @@
+      //tvstr.str() << std::endl;
+      continue;
+    }
+    if (mapAllConsumers(key)) {
```
If found to be consistent, map consumer with producer, this is "recursive" meaning anything mapped to consumer previously (running this pass backwards) is mapped to this producer as well.

csarofeen commented on torch/csrc/jit/codegen/cuda/root_domain_map.cpp:
```diff
@@ -0,0 +1,379 @@
+      std::cerr << "Equivalent Ids found: " << producer_key << " == " << pending_consumer
+                << std::endl;
+#endif
+      proveId(producer_key, pending_consumer);
```
Actually "map".

csarofeen commented on torch/csrc/jit/codegen/cuda/root_domain_map.cpp:
```diff
@@ -0,0 +1,379 @@
+  }
+}
+
+std::unordered_map<IterDomain*, IterDomain*> RootDomainMap::map(
```
Return the ID mapping of a producer-consumer.

csarofeen commented on torch/csrc/jit/codegen/cuda/utils.h:
```diff
@@ -75,6 +78,158 @@ class PolymorphicBase {
+  //! \param b An element from another equivalent class
+  //!          will create a new equivalent class if b does
+  //!          not belong to any
+  void join(T a, T b) {
```
Is this called anywhere anymore?

naoyam replied:
Yes. It's used when two objects of type `T` should be grouped into the same set. In RootDomainMap, it is used when two `IterDomain`s are mappable each other. `RootDomainMap::proveId` calls it.

csarofeen commented on torch/csrc/jit/codegen/cuda/transform_replay.cpp:
```diff
@@ -203,16 +204,52 @@ std::pair<TensorDomain*, unsigned int> TransformReplay::replayPasC(
-      TensorDomain::mapRootCtoP(consumer, producer, consumer_CA_root_ids);
+      root_map.mapConsumerToProducer(consumer, producer,
+                                     consumer_CA_root_ids);
+  if (replay_root_map != replay_root_map_old) {
```
I assume there's still some cases where this is not the same?

csarofeen reviewed (commented):
A lot of comments, but some were simply for my comprehension.

Commit:
Make ScalarCheck use const parameters

Commit:
Make individual tests as separate test functions

Commit:
Make individual tests as separate test functions

Commit:
Move DisjointSet to utils.h

Commit:
Move DisjointSet to utils.h

Commit:
Add bcast flags to BroadcastOp

Commit:
empty

Commit:
Move DisjointSet to its own header file

Commit:
Add a test for DisjointSet

Commit:
renaming

Commit:
clang-format

Commit:
Revert dropped _CUDA suffix

Commit:
Add comments

Commit:
Merge remote-tracking branch 'personal/refactoring' into expr-based-root-mapping

Commit:
Fix merge problem

Commit:
merge fix

Commit:
Merge branch '20_10_20_devel' into expr-based-root-mapping

Commit:
Minimize PR changes

Commit:
clang-format

Commit:
refactoring

Commit:
Merge branch '20_10_20_devel' into expr-based-root-mapping

Commit:
Allow broadcast domains to have different sizes depending on consumers

Commit:
test rename

Commit:
Merge branch '20_10_20_devel' into expr-based-root-mapping

Commit:
Refactored root mapping class by separating build logic into its own
class

Commit:
revert changes in the old root mapping

Commit:
Add PairwiseRootDomainMap

Commit:
Add "unsafe" root mapping

Basically the same as the previous positional mapping but just with a
small fix

Commit:
Reuse root mapping when possible

csarofeen commented on torch/csrc/jit/codegen/cuda/root_domain_map.cpp:
```diff
@@ -0,0 +1,547 @@
+  return it;
+}
+
+std::vector<Val*> getLeaves(Fusion* fusion) {
```
For the sake of mapping do we want all values? Or just those used between registered input/outputs?

naoyam replied:
Some tests use computeAt before calling addOutput. Unless we disallow that, we need to look at all values.

naoyam replied:
Let me double-check this. I wonder if that's really the case since `computeAt` itself may have problems if output values are not set as output.

naoyam replied:
There was a test that used computeAt before addOutput. Fixed it since that's invalid. Changed this part to just use `fusion->outputs()`.

Commit:
cleanup tests

Commit:
cleanup

Commit:
cleanup

Commit:
clangformat

Commit:
Remove accidentally added file

Commit:
cleanup

Commit:
cleanup

Commit:
Merge branch '20_10_20_devel' into expr-based-root-mapping

Commit:
Remove temporary change

Commit:
bug fix

Commit:
cleanup

Commit:
Merge branch '20_10_20_devel' into expr-based-root-mapping

Commit:
cleanup

tlemo commented on torch/csrc/jit/codegen/cuda/index_compute.cpp:
```diff
@@ -766,7 +767,10 @@ kir::TensorIndex* Index::getGlobalProducerIndex(
   // loop nests look like consumer
   auto producerAsC = TransformReplay::replayPasC(
-                         producer_tv->domain(), consumer_tv->domain(), -1)
+                         producer_tv->domain(),
```
can we replace this verbose parameter list with a struct?

naoyam replied:
We could, though not sure how much it would help.

tlemo commented on torch/csrc/jit/codegen/cuda/root_domain_map.h:
```diff
@@ -0,0 +1,331 @@
+class TORCH_CUDA_API UnsafePairwiseRootDomainMap
+    : public PairwiseRootDomainMap {
+ public:
+  UnsafePairwiseRootDomainMap() : PairwiseRootDomainMap() {}
```
why is this needed?

naoyam replied:
Just for backward compatibility. For correct mapping, we need to be able to look at actual expressions used to generate TensorDomains. More specifically, `BroadcastOp::is_broadcast_dims_` needs to be consulted when the expression of a producer-consumer pair is a BroadcastOp. In the current mapping, this isn't actually done, so in some cases we generate incorrect mappings. `UnsafePairwiseRootDomainMap` doesn't check the broadcast information either, so it's potentially incorrect but should be still fine if the expression is not a BroadcastOp. So, that's why it's "unsafe".

We only have one use case of this interface, which is the RFactorReplay test. In the test, TensorDomains generated with some of the replay functions are used. The use case in the test is not for general users, but it just exercises the internal interfaces for testing. Since it is a special case, I think using the unsafe interface should be reasonable.

That said, Christian said the test could be removed as we have other RFactor tests. Then, we may not need this unsafe class at all.

I don't have any specific opinion on whether we should keep the unsafe option or not.

csarofeen replied:
Is this the one that this comment belongs to?
``` .cpp
//! Creates mapping that does not consider the actual expression of
//! a producer-consumer. This is potentially unsafe.
```
Yes, let's remove this class, and remove the replay rfactor test as it's pretty outdated at this point.

naoyam replied:
Removed.

tlemo commented on torch/csrc/jit/codegen/cuda/root_domain_map.h:
```diff
@@ -0,0 +1,331 @@
+    return td() == other.td() && id() == other.id() &&
+        concreteId() == other.concreteId();
+  }
+  std::ostream& print(std::ostream& os) const;
```
the current convention is to have debug print methods simply print to `std::cout` (so, `void print() const`)

For general "stringify" operations, I'd like to encourage everyone to adopt the `toString()` pattern I introduced with `kir::toString()`

naoyam replied:
Sure. Should `toString` be defined outside of the class like `kir::toString()`?

tlemo replied:
Yes, toString() should be a free (global) function, defined in the same namespace as the type it prints (so it's found through ADL - argument depended lookup: https://en.cppreference.com/w/cpp/language/adl)

NOTE: If we use this convention, there's no need to use explicit namespaces as ADL will provide the appropriate overload (so we can just say `toString(foo)` for any type we want to "stringify")

tlemo commented on torch/csrc/jit/codegen/cuda/root_domain_map.h:
```diff
@@ -0,0 +1,331 @@
+  const IterDomain* concrete_id_ = nullptr;
+};
+
+inline std::ostream& operator<<(std::ostream& os, const DomainKey& key) {
```
`toString` is a more robust alternative to overloading `operator<<` ...

tlemo commented on torch/csrc/jit/codegen/cuda/root_domain_map.h:
```diff
@@ -0,0 +1,331 @@
+  //! \param td_alias An alias of td
+  void setAlias(const TensorDomain* td, const TensorDomain* td_alias);
+
+  std::ostream& print(std::ostream& os) const override;
```
ditto

tlemo commented on torch/csrc/jit/codegen/cuda/transform_replay.cpp:
```diff
@@ -210,9 +212,34 @@ std::pair<TensorDomain*, unsigned int> TransformReplay::replayPasC(
   }
 
-  // Map of consumer_CA_root_ids to related producer_CA_ids
   auto replay_root_map =
```
const?

naoyam replied:
Thanks. Added.

tlemo commented on torch/csrc/jit/codegen/cuda/transform_replay.cpp:
```diff
@@ -210,9 +212,34 @@ std::pair<TensorDomain*, unsigned int> TransformReplay::replayPasC(
+    // Map of consumer_CA_root_ids to related producer_CA_ids
+    std::unordered_map<IterDomain*, IterDomain*> replay_root_map_old =
+        TensorDomain::mapRootCtoP(consumer, producer, consumer_CA_root_ids);
+    if (replay_root_map != replay_root_map_old) {
```
is this temporary?

naoyam replied:
Yes. The intention is to remove this block, however, this allows us to see where divergence happens between the current mapping with the new one. I'll remove it once we finalize the PR.

csarofeen replied:
PR looks good, I think we can remove these checks now.

naoyam replied:
Removed.

tlemo reviewed (commented):
A few quick drive-by comments

Commit:
Add const

csarofeen commented on torch/csrc/jit/codegen/cuda/root_domain_map.cpp:
```diff
@@ -0,0 +1,694 @@
+    const TensorDomain* td,
+    const IterDomain* id) const {
+  DomainKey key(td, id);
+  return bcast_map_.find(key) != bcast_map_.end();
```
Isn't this the same as just checking `id->isBroadcast()`?

naoyam replied:
That's true. Changed it to `isBroadcast`.

csarofeen commented on torch/csrc/jit/codegen/cuda/root_domain_map.cpp:
```diff
@@ -0,0 +1,694 @@
+      "Not a producer-consumer pair: ",
+      producer,
+      ", ",
+      consumer);
```
What are you trying to assert above here? `fusion()->unordered_exprs()` will return all exprs in the fusion, they're not related to the producer except for the fact that they're in the same fusion. Couldn't you just do `producer->fusion() == consumer->fusion()`, then check the inputs of `consumer->getOrigin()` to see that one of them is producer?

naoyam replied:
Thanks. That was my mistake. I think I was looking at `Fusion::unordered_uses(Val*)`, but somehow I ended up writing that code. Will fix it.

csarofeen commented on torch/csrc/jit/codegen/cuda/root_domain_map.h:
```diff
@@ -0,0 +1,331 @@
+
+  //! Creates mapping that does not consider the actual expression of
+  //! a producer-consumer. This is potentially unsafe.
+  PairwiseRootDomainMap() = default;
```
Do you mean doesn't consider all expressions in the fusion? The constructor is looking at the expression containing consumer and producer.

naoyam replied:
Sorry, the comment is misleading. When `PairwiseRootDomainMap` is created with this interface, its `producer_` and `consuemr_` TensorViews are null, so the mapping won't be able to use the bcast flags, thus it's potentially unsafe. Since we don't need to have that unsafe interface, this constructor is going to be removed.

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -2206,28 +2207,86 @@ TEST(NVFuserTest, FusionBCastConcretizeRfactor_CUDA) {
+  }
+}
+
+void checkIdMapped(
```
Looks like this variant isn't used.

naoyam replied:
Removed. Thanks.

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -2206,28 +2207,86 @@ TEST(NVFuserTest, FusionBCastConcretizeRfactor_CUDA) {
 namespace {
 
-void checkIdProvedEquivalent(
+void checkIdMapped(
```
Looks like this variant isn't used.

naoyam replied:
Thanks. Removed.

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -2206,28 +2207,86 @@ TEST(NVFuserTest, FusionBCastConcretizeRfactor_CUDA) {
+      should_map);
+}
+
+void checkIdMapped(
```
Looks like this variant is always called with
```.cpp
(v0, v0->getRootDomain(), v1, v1->getRootdomain() ...
```

Should we just remove the root options on this function?

naoyam replied:
I believe I've removed the one that was not used.

csarofeen reviewed (approved):
Looks really good!

csarofeen committed:
Minor changes, minor comments.

shmsong commented on torch/csrc/jit/codegen/cuda/root_domain_map.cpp:
```diff
@@ -0,0 +1,698 @@
+void ComputeAtRootDomainMapBuilder::setMapped(
+    const DomainKey& producer,
+    const DomainKey& consumer) {
+  root_map_.eq_set_.join(producer, consumer);
```
**nitpick**: could this be wrapped in a method instead of directly modifying the states?

shmsong reviewed (approved):
This is very cool implementation. I like the way it handles reduction dependency and inconsistent mapping with a single backward pass. Left a minor nitpick but overall looks great.

Commit:
Incorporate review feedback

Commit:
Delete dead code

Commit:
Remove extra checking with the old approach

Commit:
bug fix

Commit:
Set outputs before using computeAt

Commit:
Just traverses vals from outputs because vals not contributing outputs
won't be involved in computeAt transformations.

Commit:
clang-format

Commit:
Use the toString pattern instead of printing to std::ostream

naoyam commented:
@tlemo I updated the printing logic to use `toString`. Please take a look at 68dae0ad6a94cc6d9dd6d2157ee901ddb83999f2 and let me know if you have any suggestion. Thanks!

Commit:
cleanup

Commit:
Merge branch '20_10_20_devel' into expr-based-root-mapping

Commit:
Make root mapping possible only when all concrete domains can be mapped

Fixes #446. The computeAt now throws an error as we can't compute the
broadcasted tensor at -1 without computing it twice (once for tv4 and
another for tv5).

Commit:
cleanup

tlemo commented on torch/csrc/jit/codegen/cuda/root_domain_map.h:
```diff
@@ -0,0 +1,342 @@
+  //! meaning it is a brand new domain in its TensorDomain.
+  DomainKeySet new_broadcast_domains_;
+
+  friend std::string toString(const ComputeAtRootDomainMap&);
```
nit: please move the friend declaration to the top of the struct/class definition

naoyam replied:
Thanks. Done.

Commit:
review feedback

jjsjann123 commented:
This PR is breaking broadcast before reduction. I'm marking this one and I'll open an issue shortly.
